### DISCUSSION TITLE
Why LineString can't be splitted to 2 parts？

### DISCUSSION THREAD

**Discussion by forestbat** [shapely/shapely#1781] [Q&A] [answered]
I wrote this code：
```
nearest_p = nearest_points(source_coord, origin_line)[1]
# source_coord: POINT (132.775278 45.298611)
# origin_line: LINESTRING (132.785417 45.310417, 132.772917 45.297917, 132.764583 45.297917, 132.760417 45.29375, 132.66875 45.29375)
```
And `nearest_p` is `POINT (132.7744445 45.29944449999999)`, but when I run this code:
`min_lines = split(origin_line, nearest_p)`
length of `min_lines` is 1 and in fact, `nearest_p` is not on `origin_line`.
So what happened?

**Comment by brendan-ward** [comment]
I'm pretty sure this is due to precision / robustness issues; it is difficult to produce a point that is _exactly_ on the line, so you need to use a different approach to split this.  As noted in the docstring for `split` you may need to use lower precision to help make sure the 

Note that you can also interpolate the point into the line to calculate its approximate coordinate:

```
>>> import shapely
>>> shapely.line_interpolate_point(origin_line, shapely.line_locate_point(origin_line, source_coord))
<POINT (132.774 45.299)>
```

From there, you can use an algorithm that is similar to that used by [split](https://github.com/shapely/shapely/blob/main/shapely/ops.py#L437-L457) but bypass the first check that the point intersects the line (which is returning `False` in this case).

**Reply by forestbat** [reply]
@brendan-ward Thank for your reply，but I encountered a new problem：
```
def sure_nearest_point(source_coord: Point, origin_line: LineString):
    nearest_p = nearest_points(source_coord, origin_line)[1]
    point_src = origin_line.coords[0]
    point_dst = origin_line.coords[-1]
    if ((nearest_p.x == point_src[0]) & (nearest_p.y == point_src[1])) | \
            (nearest_p.x == point_dst[0]) & (nearest_p.y == point_dst[1]):
        if len(origin_line.coords) > 2:
            rand_coord = random.randint(0, len(origin_line.coords) - 1)
            nearest_point = Point(origin_line.coords[rand_coord])
        else:
            nearest_point = Point((point_src[0] + point_dst[0]) / 2, (point_src[1] + point_dst[1]) / 2)
    else:
        nearest_point = shapely.line_interpolate_point(origin_line, shapely.line_locate_point(origin_line, source_coord))
    return nearest_point
```
```
for index in range(0, len(gpd_node_df)):
        point = gpd_node_df.geometry[index]
        near_point = sure_nearest_point(point, origin_line)
        min_lines = split_line_point_null(origin_line, near_point)
```
`split_line_point_null()` use [the algorithm](https://github.com/shapely/shapely/blob/main/shapely/ops.py#L437-L457) which you have mentioned.

But it occurs mistake:
```
……
ig_path_test.py:56: in line_min_dist
    min_lines = split_line_point_null(origin_geom_array[right], near_point)
ig_path_test.py:33: in split_line_point_null
    return [LineString(coords[: i + 2]), LineString(coords[i + 1 :])]
C:\ProgramData\Anaconda3\envs\my1stconda\lib\site-packages\shapely\geometry\linestring.py:73: in __new__
    geom = shapely.linestrings(coordinates)
C:\ProgramData\Anaconda3\envs\my1stconda\lib\site-packages\shapely\decorators.py:77: in wrapped
    return func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

coords = [[125.914583, 42.64375]], y = None, z = None, indices = None
out = None, kwargs = {}

    @multithreading_enabled
    def linestrings(coords, y=None, z=None, indices=None, out=None, **kwargs):
      ……
        if indices is None:
>           return lib.linestrings(coords, out=out, **kwargs)
E           shapely.errors.GEOSException: IllegalArgumentException: point array must contain 0 or >1 elements
C:\ProgramData\Anaconda3\envs\my1stconda\lib\site-packages\shapely\creation.py:120: GEOSException
```
I have debugged `index`es in my code, but I can't ensure what point or linestring leads to this problem, so do you have answer or tips about this crash?

**Reply by jorisvandenbossche** [reply]
Would it be possible to provide a full reproducible code example?

**Reply by forestbat** [reply]
```(python)
gpd_df_node = gpd.read_file(relpath('……'), engine='pyogrio')
gpd_df_network = gpd.read_file(abspath('……'), engine='pyogrio')


def split_line_point_null(line, splitter):
    # point is on line, get the distance from the first point on line
    distance_on_line = line.project(splitter)
    coords = list(line.coords)
    # split the line at the point and create two new lines
    current_position = 0.0
    for i in range(len(coords) - 1):
        point1 = coords[i]
        point2 = coords[i + 1]
        dx = point1[0] - point2[0]
        dy = point1[1] - point2[1]
        segment_length = (dx**2 + dy**2) ** 0.5
        current_position += segment_length
        if distance_on_line == current_position:
            # splitter is exactly on a vertex
            return [LineString(coords[: i + 2]), LineString(coords[i + 1 :])]
        elif distance_on_line < current_position:
            # splitter is between two vertices
            return [
                LineString(coords[: i + 1] + [splitter.coords[0]]),
                LineString([splitter.coords[0]] + coords[i + 1:]),
            ]
    return [line]


def line_min_dist(gpd_node_df: GeoDataFrame, gpd_line_df: GeoDataFrame):
    gpd_min_line_df = gpd.sjoin_nearest(gpd_node_df, gpd_line_df, 'left')
    gpd_min_line_df = gpd_min_line_df.loc[~gpd_min_line_df.index.duplicated()]
    index_right = gpd_min_line_df['index_right'].to_numpy()
    origin_geom_array = gpd_line_df.geometry.to_numpy()
    geom_array = np.delete(origin_geom_array, index_right)
    for index in range(0, len(gpd_node_df)):
        point = gpd_node_df.geometry[index]
        right = index_right[index]
        near_point = sure_nearest_point(point, origin_geom_array[right])
        min_lines = split_line_point_null(origin_geom_array[right], near_point)
        geom_array = np.append(geom_array, min_lines[1])
    return geom_array


def sure_nearest_point(source_coord: Point, origin_line: LineString):
    nearest_p = nearest_points(source_coord, origin_line)[1]
    point_src = origin_line.coords[0]
    point_dst = origin_line.coords[-1]
    if ((nearest_p.x == point_src[0]) & (nearest_p.y == point_src[1])) | \
            (nearest_p.x == point_dst[0]) & (nearest_p.y == point_dst[1]):
        if len(origin_line.coords) > 2:
            rand_coord = random.randint(0, len(origin_line.coords) - 1)
            nearest_point = Point(origin_line.coords[rand_coord])
        else:
            nearest_point = Point((point_src[0] + point_dst[0]) / 2, (point_src[1] + point_dst[1]) / 2)
    else:
        nearest_point = shapely.line_interpolate_point(origin_line, shapely.line_locate_point(origin_line, source_coord))
    return nearest_point


def build_graph(gpd_nodes_df, gpd_network_df, outdated: bool, cache_dir, output_path):
    geom_array = line_min_dist(gpd_nodes_df, gpd_network_df)
    geom_tree = STRtree(geom_array)
    left, right = geom_tree.query(geom_array, predicate='intersects')
    pairs = np.array([left[left != right], right[left != right]]).T
    total_graph = ig.Graph(pairs, directed=True)
    return total_graph


def test_pairs():
    print(build_graph(gpd_df_node, gpd_df_network, False, '', ''))
```

**Reply by forestbat** [reply]
@jorisvandenbossche @brendan-ward

**Reply by jorisvandenbossche** [reply]
Is it possible to share the files you are reading in (or a small subset of it)? Or manually create some data mimicking them that reproduces the issue? 
Because without actual data we still can't run your code.

**Reply by forestbat** [reply]
I'm sorry I can't log in my google account，so I can only share baidu netdisk with you：
My linestrings' shapefiles: https://pan.baidu.com/s/11sKSI0kuUjiEN22IuF-xaQ?pwd=vm32
My points' shapefiles: https://pan.baidu.com/s/1VPIvMfhIZ36H7jMb4pwk4g?pwd=3b9u

**Reply by brendan-ward** [reply]
I'm sorry @forestbat , I'm not able to figure out how to download your shapefiles from Baidu - I can't read any of the interface (sorry for cultural / language ignorance!).

If you are able to extract a small subset of your points and lines that reproduce your issue and put them in a zip file, you should be able to attach them to this issue.

**Reply by forestbat** [reply]
This is my points' file：
[points.zip](https://github.com/shapely/shapely/files/10964358/points.zip)
This is my linestrings' file：
[songliao_cut_single_new.zip](https://github.com/shapely/shapely/files/10964360/songliao_cut_single_new.zip)

**Reply by forestbat** [reply]
4 days has passed，do you reproduce this bug？
@brendan-ward @jorisvandenbossche

**Reply by brendan-ward** [reply]
@forestbat sorry for the delay.  This was not a _simple_ reproducer (though we did ask for your code!) and it involved steps that were external to splitting lines (creating a graph, etc), and steps that were non-obvious (adding a random point in one of the conditionals for `sure_nearest_point`).

However, running the core options via `geom_array = line_min_dist(gpd_nodes_df, gpd_network_df)` with the data you provided did not raise the error you indicated, it returned a set of linestrings.

I believe I found a simple reproducer, and it has to do with the split point occurring on or beyond the last point:

```
from shapely import Point, LineString

line = shapely.LineString([(0,0), (1,1)])
split_line_point_null(line, shapely.Point(1,1))

# raises
# shapely.errors.GEOSException: IllegalArgumentException: point array must contain 0 or >1 elements
```

What I think this means is that once you've found the nearest point to the line and used that to find the nearest point for that _on_ the line (`splitter`), you need to make sure that the `splitter` is _interior_ to the line - meaning that `line.project(splitter)` is >0 and < length of the line (otherwise it is exterior to the line or on the terminal coordinates).  In this case, simply don't split the line; it can be retained as-is.  Put differently, if the nearest point on the line to your original point is either the first or last vertex of the line, don't split it.

**Reply by brendan-ward** [reply]
I'm realizing I gave bad advice here, and that the implementation in shapely was already accounting for the need for the splitter to be interior to the line.  In `shapely.ops.py::_split_line_with_point`, the test `line.relate_pattern(splitter, "0********")` will only return `True` if the point is indeed interior to the line; it returns `False` if the point is either on the endpoints or not on the line, and the function then returns a single line (as expected).

Bypassing this step allowed the case where your splitter was on the endpoint or outside the line to still attempt to split the line.

I think the core problem here is that the predicates are not perfectly robust for determining relationships between points and lines, given limits of floating point precision.  That is, even after finding the nearest point on the line that the predicate testing if the point is on or interior to the line still returns False.  This isn't a bug in Shapely or GEOS, it simply is a limit of floating point math.

The way around this is to sidestep that predicate (hence my suggestion), but that means you then need to take responsibility for making sure that the splitter is indeed interior to the line before attempting to split the line (I forgot to make this very clear).  So the algorithm could go something like:

```Python
def split_line_point_null(line, splitter):
    # point is on line, get the distance from the first point on line
    distance_on_line = line.project(splitter)

    # if point is not interior to the line (between first and last coordinates)
    # or within tolerance of the endpoints, return original line
    tolerance = 1e-10
    if distance_on_line < tolerance or distance_on_line > shapely.length(line) - tolerance:
        return [line]

    # then attempt to split the line
```

You can use whatever `tolerance` is appropriate here, but helps avoid the case where your splitter is so close to the endpoints that it produces a very short line.

**Reply by forestbat** [reply]
@brendan-ward Thanks for very much，but the solution can't solve my problem completely.
I change my code like this：
```
def sure_nearest_point(source_coord: Point, origin_line: LineString):
    nearest_p = nearest_points(source_coord, origin_line)[1]
    distance_on_line = origin_line.project(source_coord)
    tolerance = 1e-5 
    point_src = origin_line.coords[0]
    point_dst = origin_line.coords[-1]
    if (((nearest_p.x == point_src[0]) & (nearest_p.y == point_src[1])) |
       ((nearest_p.x == point_dst[0]) & (nearest_p.y == point_dst[1]))) | (distance_on_line < tolerance)\
            | (distance_on_line > shapely.length(origin_line) - tolerance):
        if len(origin_line.coords) > 2:
            rand_coord = random.randint(0, len(origin_line.coords) - 1)
            nearest_point = Point(origin_line.coords[rand_coord])
        else:
            nearest_point = Point((point_src[0] + point_dst[0]) / 2, (point_src[1] + point_dst[1]) / 2)
    else:
        nearest_point = shapely.line_interpolate_point(origin_line, shapely.line_locate_point(origin_line, source_coord))
    return nearest_point
```
And there is no other change in my code，but when I use other points' shapefile which is different from `points.zip` in this discussion，the program still crashed and adjusting `tolerance` has no effect.

**Reply by brendan-ward** [reply]
The `tolerance` parameter I suggested is for determining if you should split the line _after_ finding the nearest point, not for determining the nearest point.

**Reply by forestbat** [reply]
I wrote a test for ensuring tolerance：
```(python)
def test_tolerance():
    gpd_min_line_df = gpd.sjoin_nearest(gpd_df_node, gpd_df_network, 'left')
    gpd_min_line_df = gpd_min_line_df.loc[~gpd_min_line_df.index.duplicated()]
    index_right = gpd_min_line_df['index_right'].to_numpy()
    origin_geom_array = gpd_df_network.geometry.to_numpy()
    project_list = []
    for index in range(0, len(gpd_df_node)):
        point = gpd_df_node.geometry[index]
        right = index_right[index]
        near_point = sure_nearest_point_old(point, origin_geom_array[right])
        project_list.append(origin_geom_array[right].project(near_point))
    print(project_list)

def sure_nearest_point_old(source_coord: Point, origin_line: LineString):
    nearest_p = nearest_points(source_coord, origin_line)[1]
    point_src = origin_line.coords[0]
    point_dst = origin_line.coords[-1]
    if ((nearest_p.x == point_src[0]) & (nearest_p.y == point_src[1])) | \
            (nearest_p.x == point_dst[0]) & (nearest_p.y == point_dst[1]):
        if len(origin_line.coords) > 2:
            rand_coord = random.randint(0, len(origin_line.coords) - 1)
            nearest_point = Point(origin_line.coords[rand_coord])
        else:
            nearest_point = Point((point_src[0] + point_dst[0]) / 2, (point_src[1] + point_dst[1]) / 2)
    else:
        nearest_point = shapely.line_interpolate_point(origin_line, shapely.line_locate_point(origin_line, source_coord))
    return nearest_point
```
And project_list looks like this：
```
[0.04432260408057233, 0.04218472542212115, 0.0637690182941307, 0.004222999999996092, 0.015517458313150477, 0.09440157625101754, 0.0003060000000019158, 0.0772293487644716, 0.05003792933941874, 0.025677669529667435, 0.032753191317823255, 0.08558036702676952, 0.0105829999999969, 0.10315567835199253, 0.06703041916308822, 0.019416231165248918, 0.006528000000002976, 0.06037599046457332, 0.018421027914405343, 0.010027000000000898, 0.02226096246530259, 0.05426673554460159, 0.047198147565396996, 0.03337196246530016, 0.06307111799533686, ……]
```
You can realize 0.04 degree is too big to "ignore" them in a tolerance, so what's the problem and how to solve?

**Reply by brendan-ward** [reply]
I don't understand why you are doing this block:

```Python
    if ((nearest_p.x == point_src[0]) & (nearest_p.y == point_src[1])) | \
            (nearest_p.x == point_dst[0]) & (nearest_p.y == point_dst[1]):
        if len(origin_line.coords) > 2:
            rand_coord = random.randint(0, len(origin_line.coords) - 1)
            nearest_point = Point(origin_line.coords[rand_coord])
        else:
            nearest_point = Point((point_src[0] + point_dst[0]) / 2, (point_src[1] + point_dst[1]) / 2)
```

which, if I'm interpreting it correctly, is basically: if the nearest point is the first or last point of the line and the line has more than 2 points, pick a random vertex on that line, and if the line only has 2 points, use the midpoint on the line

Both of which offset your nearest point from the actual nearest point.  If the nearest point on the line to your original point is an endpoint (first or last point) on the target line, there is no reason to split the line: the line is already split at that point!

Instead, I think you could replace the call to `sure_nearest_point` with

```
near_point = shapely.line_interpolate_point(
    origin_geom_array[right], shapely.line_locate_point(origin_geom_array[right], point)
)
```

Because that also finds the nearest coordinate.

This might make it a little more clear:

```Python
point = Point(0,0)
line = LineString([(1,1), (2,2)])

shapely.line_interpolate_point(line, shapely.line_locate_point(line, point))
# <POINT (1 1)>

point = Point(1.5,1.25)
shapely.line_interpolate_point(line, shapely.line_locate_point(line, point))
# <POINT (1.375 1.375)>
```

Using this nearest point, you then need to decide if you are going to split the line or if it is already effectively split because the nearest point to your original point is the endpoint of the line.  Since testing equality of floating point numbers is problematic, _one_ way to approach this is to calculate the position on the line and use a tolerance to make the decision:

```Python
tolerance = 1e-10

nearest_point = shapely.line_interpolate_point(line, shapely.line_locate_point(line, point))

pos = shapely.line_locate_point(line, nearest_point)
if pos > tolerance and pos < shapely.length(line) - tolerance:
    # split the line...
else:
    # use the original line
```

This ensures that you only split the line, and add a new coordinate, if that new coordinate is _between_ the endpoints of the line.

If you wanted, you could make this more complex by checking to see if the nearest point is within a tolerance of one of the existing vertices on the line other than the endpoint.  E.g., if your line has 3 vertices and the point is very very close to the middle vertex, then reuse that existing vertex as the split point instead.


If you are doing this for many lines and points, it would benefit from taking a vectorized approach - but I'd recommend getting a solid handle on the pure python approach first.  Some ideas at: https://github.com/pygeos/pygeos/issues/352

Some functionality in this space has been added to GEOS [recently](https://github.com/libgeos/geos/issues/706) but I don't think that has been implemented in Shapely yet.

**Reply by forestbat** [reply]
Yes now I replaced my code with your method：
```
def test_tolerance():
    gpd_min_line_df = gpd.sjoin_nearest(gpd_df_node, gpd_df_network, 'left')
    gpd_min_line_df = gpd_min_line_df.loc[~gpd_min_line_df.index.duplicated()]
    index_right = gpd_min_line_df['index_right'].to_numpy()
    origin_geom_array = gpd_df_network.geometry.to_numpy()
    project_list = []
    # Time complexity is O(m), m is amount of points
    for index in range(0, len(gpd_df_node)):
        point = gpd_df_node.geometry[index]
        right = index_right[index]
        near_point = shapely.line_interpolate_point(origin_geom_array[right],
                                                    shapely.line_locate_point(origin_geom_array[right], point))
        project_list.append(origin_geom_array[right].project(near_point))
    print(project_list)
```
But project list is still like this, too big:
```
0.04432260408057233, 0.04218472542212115, 0.0637690182941307, 0.004222999999996092, 0.015517458313150477, 0.09440157625101754, 0.0003060000000019158, 0.0772293487644716, 0.05003792933941874, 0.025677669529667435, 0.032753191317823255, 0.08558036702676952, 0.0105829999999969, 0.10315567835199253, 0.06703041916308822, 0.019416231165248918, 0.006528000000002976……
```
My algorithm to find nearest lines comes from [here](https://github.com/geopandas/geopandas/discussions/2773).

**Comment by forestbat** [comment, marked as answer]
Now I use a new method to solve this problem：
```(python)
def line_min_dist(gpd_node_df: GeoDataFrame, gpd_line_df: GeoDataFrame):
    gpd_min_line_df = gpd.sjoin_nearest(gpd_node_df, gpd_line_df, 'left')
    gpd_min_line_df = gpd_min_line_df.loc[~gpd_min_line_df.index.duplicated()]
    index_right = gpd_min_line_df['index_right'].to_numpy()
    origin_geom_array = gpd_line_df.geometry.to_numpy()
    geom_array = np.delete(origin_geom_array, index_right)
    for index in range(0, len(gpd_node_df)):
        point = gpd_node_df.geometry[index]
        right = index_right[index]
        pre_near_point = sure_nearest_point(point, origin_geom_array[right])
        line = LineString([point, pre_near_point])
        if line.length != 0:
            ex_line = get_extrapolated_line(point, pre_near_point, 1 / line.length)
            min_lines = split(origin_geom_array[right], ex_line)
        else:
            min_lines = split(origin_geom_array[right], pre_near_point)
        geom_array = np.append(geom_array, (list(min_lines.geoms))[1])
    return geom_array


def sure_nearest_point(source_coord: Point, origin_line: LineString):
    nearest_p = nearest_points(source_coord, origin_line)[1]
    point_src = origin_line.coords[0]
    point_dst = origin_line.coords[-1]
    point_neigh = origin_line.coords[1]
    if ((nearest_p.x == point_src[0]) & (nearest_p.y == point_src[1])) | ((nearest_p.x == point_dst[0])
                                                                          & (nearest_p.y == point_dst[1])) | \
            ((source_coord.x == nearest_p.x) & (source_coord.y == nearest_p.y)):
        nearest_point = Point((point_src[0] + point_neigh[0]) / 2, (point_src[1] + point_neigh[1]) / 2)
    else:
        nearest_point = shapely.line_interpolate_point(origin_line,
                                                       shapely.line_locate_point(origin_line, source_coord))
    return nearest_point


def get_extrapolated_line(source_coord, coord, extrapolate_ratio):
    source = Point(source_coord)
    p2 = Point(coord)
    c = (source.x + extrapolate_ratio * (p2.x - source.x), source.y + extrapolate_ratio * (p2.y - source.y))
    return LineString([source, c])
```
